sled-agent/src/instance.rs

```diff
@@ -186,9 +198,117 @@ struct PropolisSetup {
+        tx: oneshot::Sender<Result<SledInstanceState, ManagerError>>,
+    },
+    Terminate {
+        tx: oneshot::Sender<Result<InstanceUnregisterResponse, ManagerError>>,
+    },
+    IssueSnapshotRequest {
+        disk_id: Uuid,
+        snapshot_id: Uuid,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    AddExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+    DeleteExternalIp {
+        ip: InstanceExternalIpBody,
+        tx: oneshot::Sender<Result<(), ManagerError>>,
+    },
+}
+
+// A small task which tracks the state of the instance, by constantly querying
+// the state of Propolis for updates.
+//
+// This task communicates with the "InstanceRunner" task to report status.
+struct InstanceMonitorRunner {
+    client: Arc<PropolisClient>,
+    tx_monitor: mpsc::Sender<InstanceMonitorRequest>,
+}
+
+impl InstanceMonitorRunner {
+    async fn run(self) -> Result<(), anyhow::Error> {
+        let mut gen = 0;
+        loop {
+            // State monitoring always returns the most recent state/gen pair
+            // known to Propolis.
+            let response = self
+                .client
+                .instance_state_monitor()
+                .body(propolis_client::types::InstanceStateMonitorRequest {
+                    gen,
+                })
+                .send()
+                .await
+                .map_err(|e| anyhow!(e))?
+                .into_inner();
+            let observed_gen = response.gen;
+
+            // Now that we have the response from Propolis' HTTP server, we
+            // forward that to the InstanceRunner.
+            //
+            // It will decide the new state, provide that info to Nexus,
+            // and possibly identify if we should terminate.
+            let (tx, rx) = oneshot::channel();
+            self.tx_monitor
+                .send(InstanceMonitorRequest::Update { state: response, tx })
+                .await
+                .map_err(|e| anyhow!(e))?;
+
+            if let Reaction::Terminate = rx.await.map_err(|e| anyhow!(e))? {
+                return Ok(());
+            }
+
+            // Update the generation number we're asking for, to ensure the
+            // Propolis will only return more recent values.
+            gen = observed_gen + 1;
+        }
+    }
+}
+
+enum InstanceMonitorRequest {
+    Update {
+        state: propolis_client::types::InstanceStateMonitorResponse,
+        tx: oneshot::Sender<Reaction>,
+    },
+}
+
+struct InstanceRunner {
     log: Logger,
 
+    // A oneshot which, when closed, will cause the "run" loop to terminate
+    // without processing any subsequent messages.
+    //
+    // We need the tx side of this thing to be alive to prevent "terminate_rx"
+    // from shutting down early.
+    #[allow(dead_code)]
+    terminate_tx: oneshot::Sender<()>,
+    terminate_rx: oneshot::Receiver<()>,
```

Comment:
I swear there must be a simpler solution here, but this is basically trying to cover the need for:

- I have a "runner" object, which is basically a loop + `tokio::select`
- I want to let the runner terminate itself . No one else needs this ability externally.

It feels like it must be possible to do this with a smaller / more lightweight object, but both ends of the oneshot also seemed very straightforward.

Reply:
A `oneshot` is [actually quite lightweight](https://docs.rs/tokio/latest/src/tokio/sync/oneshot.rs.html)  The other option in `tokio::sync` would be an `Arc<tokio::sync::Notify>`, which can be used to wake a task _without_ sending data, but since a `Notify` can be used to wake a task multiple times, the implementation is actually much more complex.

In this specific case, though, since the termination is only sent from "inside" the task, you don't really need to wake up the task --- the task is (AFAICT) always running when it terminates itself. So, as far as I can tell, you should be able to just replace this with `terminating: bool`, and then just put
```rust
if self.terminating {
    return Ok(());
}
```
at the top of the `loop` in `InstanceRunner::run`, before the `tokio::select!`. You'll hit that every time the loop iterates, before you enter the `select!` and possibly yield.

Reply:
...actually, rather than writing `if self.terminating`, you could just change the loop to `while !self.terminating`.

Reply:
Ah, that's a good point! I used a `oneshot` to make sure I wouldn't miss this as a "wakeup call", but it's true that if a task is only flagging this for itself, this can be much simpler.

Reply:
Although, hmm. It is worth noting that if the terminating oneshot is sent, and then the task `.await`s something else that isn't ready, it will have woken itself and the `select!` will fall through into the terminating branch. This isn't the case with a `bool`. So, if you represent termination with a bool, set the flag, and then `.await` something, you're not going to "see" the termination until the thing you await wakes you.

If we don't `.await` anything after setting the flag, that's not necessarily an issue. But, it's worth calling out...

Reply:
Just to make sure I'm grokking this:

- If we use a boolean that we check before the `select` call, the `InstanceRunner` can self-terminate, but only after one of the select arms has "fully completed"
- If we use a oneshot, we can bail-out before any other select arms "fully complete"

For example, something like:


```rust
while !self.terminating {
  tokio::select! {
    let a = self.foo() => ...
    let b = self.bar() => ...
  }
}

async fn foo(&mut self) {
  do_stuff.await;

  self.terminating = true;

  // Other `select` arms could continue being queried concurrently while we're awaiting this.
  // If we closed a oneshot, however, we'd exit before getting to query this.
  do_stuff.await;
}
```

Reply:
@smklein yup, that's correct.

Reply:
If that's the case, I *think* the boolean approach here is okay. We will only set this value on ourselves, and completing "our dispatched operation" is okay. We *do* want to e.g. post our state to Nexus if we can, before terminating the runner, which is another asynchronous operation.